src/hooks/execution/SpendingLimitHook.sol

```diff
@@ -0,0 +1,240 @@
+}
+
+/// @title SpendingLimitHook
+/// @notice Example use case of an execution hook to set spending limits on specific key hashes.
+/// @author modified from https://github.com/ithacaxyz/account/blob/main/src/GuardedExecutor.sol
+contract SpendingLimitHook is ISpendingLimitHook {
+    using EnumerableSetLib for *;
+    using DynamicArrayLib for *;
+    using AccountKeyHashLib for bytes32;
+
+    ////////////////////////////////////////////////////////////////////////
+    // Structs
+    ////////////////////////////////////////////////////////////////////////
+
+    /// @dev Information about a spend.
+    /// All timestamp related values are Unix timestamps in seconds.
+    struct SpendInfo {
+        /// @dev Address of the token. `address(0)` denotes native token.
+        address token;
+        /// @dev The type of period.
+        SpendPeriod period;
+        /// @dev The maximum spend limit for the period.
+        uint256 limit;
+        /// @dev The amount spent in the last updated period.
+        uint256 spent;
+        /// @dev The start of the last updated period.
+        uint256 lastUpdated;
+        /// @dev The amount spent in the current period.
+        uint256 currentSpent;
+        /// @dev The start of the current period.
+        uint256 current;
+    }
+
+    /// @dev Holds the storage for the token period spend limits.
+    /// All timestamp related values are Unix timestamps in seconds.
+    struct TokenPeriodSpendStorage {
+        /// @dev The maximum spend limit for the period.
+        uint256 limit;
+        /// @dev The amount spent in the last updated period.
+        uint256 spent;
+        /// @dev The start of the last updated period (unix timestamp).
+        uint256 lastUpdated;
+    }
+
+    /// @dev Holds the storage for the token spend limits.
+    struct TokenSpendStorage {
+        /// @dev An enumerable set of the periods.
+        EnumerableSetLib.Uint8Set periods;
+        /// @dev Mapping of `uint8(period)` to `TokenPeriodSpendStorage`.
+        mapping(uint256 => TokenPeriodSpendStorage) spends;
+    }
+
+    /// @dev Holds the storage for spend permissions and the current spend state.
+    struct SpendStorage {
+        /// @dev An enumerable set of the tokens.
+        EnumerableSetLib.AddressSet tokens;
+        /// @dev Mapping of `token` to `TokenSpendStorage`.
+        mapping(address => TokenSpendStorage) spends;
+    }
+
+    mapping(AccountKeyHash => SpendStorage) public spendStorage;
+
+    struct TempStorage {
+        uint256 totalNativeSpend;
+        address[] erc20s;
+        uint256[] transferAmounts;
+        uint256[] balancesBefore;
+    }
+
+    /// @dev Set the spend limit for a key hash.
+    /// Uses msg.sender to compute the accountKeyHash.
+    function setSpendLimit(bytes32 keyHash, address token, SpendPeriod period, uint256 limit) external {
+        AccountKeyHash accountKeyHash = keyHash.wrap(msg.sender);
```

Comment:
maybe use a different word? wrap is keyword for a custom type

Reply:
ok! I did `hashSender`